sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala

```diff
@@ -607,6 +607,89 @@ object ScalaReflection extends ScalaReflection {
       }
     }
   }
+
+  /**
+   * Validates a type for encoder recursively.  If it's not supported, i.e. there is no encoder can
+   * be built for this type, an [[UnsupportedOperationException]] will be thrown with detailed
+   * error message to explain the type path walked so far and which class we are not supporting.
+   * There are 4 kinds of type path:
+   *  * the root type: `root class: "abc.xyz.MyClass"`
+   *  * the value type of [[Option]]: `option value class: "abc.xyz.MyClass"`
+   *  * the element type of [[Array]] or [[Seq]]: `array element class: "abc.xyz.MyClass"`
+   *  * the field of [[Product]]: `field (class: "abc.xyz.MyClass", name: "myField")`
+   */
+  def validateTypeForEncoder(tpe: `Type`): Unit = {
+    val clsName = getClassNameFromType(tpe)
+    validateTypeForEncoder(tpe, s"""- root class: "${clsName}"""" :: Nil)
+  }
+
+  private def validateTypeForEncoder(tpe: `Type`, walkedTypePath: Seq[String]): Unit =
```

Comment:
Why is this a separate function now?  Now we have duplicated copies of the same logic that can fall out of sync.